File: ci-tests/tests/testers/ignore_filter.py
from .hoard_tester import HoardTester
from pathlib import Path
import os


class IgnoreFilterTester(HoardTester):
    def __init__(self):
        self.reset()
        # We are not changing env on this test
        self.env = {"USE_ENV": "1"}
        # All three of global, hoard, and pile-ignore files should be ignored.
        self.global_file = "global_ignore"
        self.hoard_file = "ignore_for_hoard"
        self.pile_file = "spilem"
        self.nested_file = "nested_dir/.hidden"

        # Create files to ignore
        self.anon_dir_root = Path.home().joinpath("first_anon_dir")
        self.named_dir1_root = Path.home().joinpath("first_named_dir1")
        self.named_dir2_root = Path.home().joinpath("first_named_dir2")

        for root in [self.anon_dir_root, self.named_dir1_root, self.named_dir2_root]:
            for file in [self.global_file, self.hoard_file, self.pile_file, self.nested_file]:
                self.generate_file(root.joinpath(file))

    def run_test(self):
        # Run hoard
        self.run_hoard("backup")
        # Run again to be sure
        self.run_hoard("backup")

        # Delete unexpected files for assert_same_tree
        # Named dir1 pile should ignore all
        for file in [self.global_file, self.hoard_file, self.pile_file]:
            os.remove(self.named_dir1_root.joinpath(file))
        # Named dir2 pile should only ignore hoard and global
        for file in [self.global_file, self.hoard_file]:
            os.remove(self.named_dir2_root.joinpath(file))
        # Anon dir should only ignore global
        os.remove(self.anon_dir_root.joinpath(self.global_file))

        # Assert trees
        data_dir = self.data_dir_path()
        self.assert_same_tree(
            self.anon_dir_root,
            data_dir.joinpath("hoards", "anon_dir"),
        )
        self.assert_same_tree(
            self.named_dir1_root,
            data_dir.joinpath("hoards", "named", "dir1"),
        )
        self.assert_same_tree(
            self.named_dir2_root,
            data_dir.joinpath("hoards", "named", "dir2"),
        )

File: tests/ignore_filter.rs
mod common;

use common::base::DefaultConfigTester;
use hoard::command::Command;
use tokio::fs;
use std::path::PathBuf;

const GLOBAL_FILE: &str = "global_ignore";
const HOARD_FILE: &str = "ignore_for_hoard";
const PILE_FILE: &str = "spilem";
const NESTED_FILE: &str = "nested_dir/.hidden";

fn ignored_files(tester: &DefaultConfigTester) -> Vec<PathBuf> {
    vec![
        tester.home_dir().join("first_anon_dir").join(GLOBAL_FILE),
        tester.home_dir().join("first_named_dir1").join(GLOBAL_FILE),
        tester.home_dir().join("first_named_dir2").join(GLOBAL_FILE),
        tester.home_dir().join("first_named_dir1").join(HOARD_FILE),
        tester.home_dir().join("first_named_dir2").join(HOARD_FILE),
        tester.home_dir().join("first_named_dir1").join(PILE_FILE),
        tester.home_dir().join("first_named_dir2").join(NESTED_FILE),
    ]
}

fn all_extra_files(tester: &DefaultConfigTester) -> Vec<PathBuf> {
    ["first_anon_dir", "first_named_dir1", "first_named_dir2"]
        .into_iter()
        .flat_map(|slug| {
            vec![
                tester.home_dir().join(slug).join("global_ignore"),
                tester.home_dir().join(slug).join("ignore_for_hoard"),
                tester.home_dir().join(slug).join("spilem"),
                tester
                    .home_dir()
                    .join(slug)
                    .join("nested_dir")
                    .join(".hidden"),
            ]
        })
        .collect()
}

#[tokio::test]
async fn test_ignore_filter() {
    let mut tester = DefaultConfigTester::new().await;
    tester.setup_files().await;
    tester.use_first_env();

    for home in all_extra_files(&tester) {
        common::create_file_with_random_data::<2048>(&home).await;
    }

    tester.expect_command(Command::Backup { hoards: Vec::new() }).await;

    // Delete ignored files from home so assertion works
    for home in ignored_files(&tester) {
        fs::remove_file(&home).await.expect("failed to remove ignored file");
    }

    tester.assert_first_tree().await;
}
